subprojects/gradle/critical-path/api/src/main/kotlin/com/avito/android/critical_path/CriticalPathRegistry.kt

```diff
@@ -0,0 +1,32 @@
+package com.avito.android.critical_path
+
+import com.avito.android.critical_path.internal.PathBuildProvider
+import com.avito.android.gradle.metric.GradleCollector
+import com.avito.kotlin.dsl.isRoot
+import org.gradle.api.Project
+
+public object CriticalPathRegistry {
+
+    public fun addListener(project: Project, listener: CriticalPathListener) {
+        check(project.isRoot()) {
+            "Can consume build critical path only in root project but was ${project.path}"
+        }
+        getProvider(project).addPathListener(listener)
+    }
+
+    private fun getProvider(project: Project): PathBuildProvider {
+        return if (project.extensions.extraProperties.has(listenerExtra)) {
+            project.extensions.extraProperties.get(listenerExtra) as PathBuildProvider
+        } else {
+            val buildListener = PathBuildProvider()
+
+            GradleCollector.initialize(project, listOf(buildListener))
```

Comment:
Do we have some requirements where should call `addListener`?

Reply:
Before the end of a build